impl/src/as/mod.rs

```diff
@@ -0,0 +1,371 @@
+//! Implementations of [`AsRef`]/[`AsMut`] derive macros.
+
+#[cfg(feature = "as_mut")]
+pub(crate) mod r#mut;
+#[cfg(feature = "as_ref")]
+pub(crate) mod r#ref;
+
+use proc_macro2::TokenStream;
+use quote::quote;
+use syn::{
+    parse::{Parse, ParseStream},
+    spanned::Spanned,
+    Token,
+};
+
+use crate::utils::{add_extra_generic_type_param, add_extra_where_clauses, Either};
+
+/// Type alias for an expansion context:
+/// - Derived trait [`syn::Ident`]
+/// - Derived trait method [`syn::Ident`]
+/// - Type name used as generic in blanket impls [`syn::Ident`]
+/// - Optional `mut` token used in expansion
+type ExpansionCtx<'a> = (
+    &'a syn::Ident,
+    &'a syn::Ident,
+    &'a syn::Ident,
+    Option<&'a Token![mut]>,
+);
+
+/// Expands a [`AsRef`]/[`AsMut`] derive macro.
+pub fn expand(
+    input: &syn::DeriveInput,
+    (trait_ident, method_ident, conv_type, mutability): ExpansionCtx<'_>,
+) -> syn::Result<TokenStream> {
+    let trait_path = quote! { ::derive_more::#trait_ident };
+
+    let field_args = extract_field_args(input, trait_ident, method_ident)?;
+    let (impl_generics, ty_generics, where_clause) = input.generics.split_for_impl();
+    let input_type = &input.ident;
+
+    Ok(
+        field_args
+            .into_iter()
+            .map(
+                |FieldArgs {forward, field, ident}| {
+                    let member = quote! { self.#ident };
+                    let field_type = &field.ty;
+                    let (
+                        body,
+                        impl_generics,
+                        where_clause,
+                        trait_path,
+                        return_type
+                    ) = if forward {
+                        let trait_path = quote! { #trait_path<#conv_type> };
```

Comment:
In general I prefer to have a bit less nested code than this. If it's reasonably possible to un-indent this somehow I would like that. Feel free to ignore this comment though if it requires major changes.